File: lintcode/007-Binary-Tree-Serialization/solution.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * This method will be invoked first, you should design your own algorithm
     * to serialize a binary tree which denote by a root node to a string which
     * can be easily deserialized by your own "deserialize" method later.
     */
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder("");
        if (root == null) {
            return sb.toString();
        }
        // do dfs to build tree, use "#" to represent null
        Stack<TreeNode> stack = new Stack<>();
        stack.add(root);
        while (!stack.empty()) {
            TreeNode curr = stack.pop();
            if (curr == null) {
                sb.append("#,");
            } else {
                sb.append(curr.val).append(",");
                stack.add(curr.right);
                stack.add(curr.left);
            }
        }
        // remove last comma
        sb.deleteCharAt(sb.length() - 1);
        return sb.toString();
    }

    /**
     * This method will be invoked second, the argument data is what exactly
     * you serialized at method "serialize", that means the data is not given by
     * system, it's given by your own serialize method. So the format of data is
     * designed by yourself, and deserialize it here as you serialize it in
     * "serialize" method.
     */
    public TreeNode deserialize(String data) {
        if (data == null || data.length() == 0) {
            return null;
        }
        
        StringTokenizer st = new StringTokenizer(data, ",");
        return readTree(st);
    }
    
    public TreeNode readTree(StringTokenizer st) {
        if (!st.hasMoreTokens()) {
            return null;
        }
        
        String val = st.nextToken();
        if (val.equals("#")) {
            return null;
        }
        TreeNode root = new TreeNode(Integer.parseInt(val));
        root.left = readTree(st);
        root.right = readTree(st);
        
        return root;
    }
}

File: lintcode/007-Binary-Tree-Serialization/solution.cpp
class Solution {
public:
    string serialize(TreeNode *root) {
        string result;
        writeBinaryTree(root, result);
        
        return result;
    }

    TreeNode *deserialize(string data) {
        stringstream ss(data);
        TreeNode *root;
        readBinaryTree(root, ss);

        return root;
    }
private:
    // {1,#,3,2,1,#,#,1,#}
    void writeBinaryTree(TreeNode *root, string &result) {
        if (root == nullptr) {
            result += "#,";
        } else {
            result += to_string(root->val) + ",";
            writeBinaryTree(root->left, result);
            writeBinaryTree(root->right, result);
        }
    }

    void readBinaryTree(TreeNode* &root, stringstream &ss) {
        string curr;
        if (!getline(ss, curr, ',')) {
            return;
        }
        if (curr != "#") {
            root = new TreeNode(stoi(curr));
            readBinaryTree(root->left, ss);
            readBinaryTree(root->right, ss);
        } else {
            root = nullptr;
        }
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (root == nullptr) return "";
        string s;
        stack<TreeNode*> nodes;
        nodes.push(root);
        while (!nodes.empty()) {
            TreeNode *curr = nodes.top();
            nodes.pop();
            if (curr == nullptr) {
                s.append("#,");
            } else {
                s.append(to_string(curr->val) + ",");
                nodes.push(curr->right);
                nodes.push(curr->left);
            }
        }
        
        // remove last `,`
        s.resize(s.size() - 1);
        cout << s << endl;
        return s;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data.size() == 0) return nullptr;
        stringstream ss(data);
        string val;
        TreeNode *root;
        stack<TreeNode**> nodes;
        nodes.push(&root);
        while (!nodes.empty()) {
            TreeNode **curr = nodes.top();
            nodes.pop();
            
            if (!getline(ss, val, ',')) {
                break;
            }
            
            if (val == "#") {
                *curr = nullptr;
            } else {
                *curr = new TreeNode(stoi(val));
                nodes.push(&(*curr)->right);
                nodes.push(&(*curr)->left);
            }
        }
        
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));